orderable/models.py

```diff
@@ -101,25 +80,53 @@ def _move_to_new_pos():
             #     sort_order > new_pos.
             to_shift = to_shift.filter(sort_order__gte=self.sort_order)
             to_shift.update(sort_order=models.F('sort_order') + 1)
-            _move_to_new_pos()
+            self.sort_order = new_pos
 
         # self.sort_order decreased.
         elif old_pos and new_pos < old_pos:
-            _move_to_end()
+            self._move_to_end(objects)
+            super(Orderable, self).save()
             # Increment `sort_order` on objects with:
             #     sort_order >= new_pos and sort_order < old_pos
             to_shift = to_shift.filter(sort_order__gte=new_pos, sort_order__lt=old_pos)
             to_shift.update(sort_order=models.F('sort_order') + 1)
-            _move_to_new_pos()
+            self.sort_order = new_pos
 
         # self.sort_order increased.
         elif old_pos and new_pos > old_pos:
-            _move_to_end()
+            self._move_to_end(objects)
+            super(Orderable, self).save()
             # Decrement sort_order on objects with:
             #     sort_order <= new_pos and sort_order > old_pos.
             to_shift = to_shift.filter(sort_order__lte=new_pos, sort_order__gt=old_pos)
             to_shift.update(sort_order=models.F('sort_order') - 1)
-            _move_to_new_pos()
+            self.sort_order = new_pos
+
+    def _move_to_end(self, objects):
+        """Temporarily save `self.sort_order` elsewhere (max_obj)."""
+        max_obj = objects.all().aggregate(models.Max('sort_order'))['sort_order__max']
+        self.sort_order = max_obj + 1 if max_obj else 1
+
+    def save(self, *args, **kwargs):
+        """
+        Keep the unique order in sync.
+
+        Expects to be run in a transaction to avoid race conditions.
+
+        WARNING: Intensive giggery-pokery zone.
+        """
+        objects = self.get_filtered_manager()
+        old_pos = getattr(self, '_original_sort_order', None)
+        new_pos = self.sort_order
+
+        try:
+            with transaction.atomic():
+                self._save(objects, old_pos, new_pos)
+        except IntegrityError:
+            with transaction.atomic():
+                old_pos = objects.filter(pk=self.pk).values_list(
+                    'sort_order', flat=True)[0]
```

Comment:
It's basically the second test case in the `SubTasks` section - that fails without this change. The basic problem is that the `old_pos` can be changed by the saving of another object without the object here being reloaded - "formset" mode. This happens when using both `OrderableTabularInline` and the `reorder` view on the list page. You won't see it when not using `unique_together` as there is eventual uniqueness of the ids.

More detail:
Suppose 4 objects (ABCD) with sort orders 1-4. Load objects AB from the database, and set them to have sort orders 3 and 4 respectively and save them. The following happens:
- A is moved out of the way (position 5)
- BC are moved down one place (B to 1, C to 2)
- A is moved back to position 3

So far so good. We now have BCAD. Now we save B with sort order 4, but it has `old_pos=2` because it has not been loaded from the database. So:
- B is moved out of the way (position 5)
- A is moved down one place (as it is in position 3). This clashes with the new position of C and throws an integrity error.
- (Now) we catch that, recheck the current position of B and move everything as normal.

The whole process is still imperfect at this level as we now have CADB which isn't what we asked for, kinda. However in the context where multiple values are assigned, it is most likely we are assigning all four items to new places as a reordering view. i.e. C to position 1 and D to position 2.

Saving C (from position 3 to position 1) we would try to move positions 1 and 2 up, but of course D is now in C's position so this fails as well and we recheck that C is in fact in the correct place and we have nothing to do.

Saving D (from position 4 to position 2) would not hit an error in this case as only A would be moved (back to its correct position 3) and D would go 3-5-2. We then have CDAB as requested.

I don't know if there is a permutation which can result in the wrong order after all of these have been applied - it is quite possible but I can't be bothered to work out the maths to try and prove it. It would be better if there was an API for explicitly setting the ordering which was used by the relevant places in the admin. This is one of the things I want to add.